### PR TITLE
Ensure all specs are validated

### LINKED ISSUE #507: Ensure all specs are validated
## Description

Currently, there are a few edge cases where specs do not get validated, as raised in https://github.com/hdmf-dev/hdmf/pull/500. We want to ensure that all specs get correctly validated.

1. if a group includes two or more children with the same defined data type but a different name (independent of the quantity), then only the last instance will be validated. Eg. for an electrodes group that has VectorData X, Y, and Z - only Z will be validated. The validator should validate each of these children.
2. if a child link has the same data type as a child group or dataset, then the link will be validated but the group or dataset with the same data type will not be validated. The validator should validate both the link and the group or dataset.
3. if a group specifies two unnamed child groups of type A and B, where B inherits from A, then a single child of type B satisfies the spec for both B and A. The validator should require distinct children to match each spec. Both of the following cases should satisfy the spec: a) one child of type A and one of type B, or b) two children of type B.

## Steps to Reproduce

Create an hdmf object and spec which should be invalid according to one of the three points above, and see that it passes validation.

## Environment

    Python Executable: Conda or Python
    Python Version: Python 3.5, 3.6, 3.7, or 3.8
    Operating System: Windows, macOS or Linux
    HDMF Version: 2.3.0

## Checklist

- [x] Have you ensured the bug was not already [reported](https://github.com/hdmf-dev/hdmf/issues) ?
- [x] Have you included a brief and descriptive title?
- [x] Have you included a clear description of the problem you are trying to solve?
- [ ] Have you included a minimal code snippet that reproduces the issue you are encountering?
- [x] Have you checked our [Contributing](https://github.com/hdmf-dev/hdmf/blob/dev/docs/CONTRIBUTING.rst) document?

### PR DESCRIPTION
**Pull request by dsleiter**
## Motivation

* Fix #507
* Significant refactoring of the GroupValidator was done in an attempt to make it more easily understandable and facilitate future development
* The logic for matching group builder children against group spec children was extracted to a new SpecMatcher class in order to isolate the matching algorithm, which will need to be improved in future iterations
* New unit tests were added for the improvements listed in #507
* Refactoring was done with future validator issues in mind, such as raising a warning for superfluous builders attached to a group (https://github.com/NeurodataWithoutBorders/pynwb/issues/1332 and https://github.com/NeurodataWithoutBorders/pynwb/issues/942)

I realize that this is a pretty significant change to the GroupValidator code, and am very open to critical comments. Do you think the changes improve readability enough to warrant the change? Is there any clear misunderstanding of how validation should be done?

## How to test the behavior?

Execute the unit tests. All previously existing tests still pass after refactoring, in addition to the new tests.

## Checklist

- [x] Did you update CHANGELOG.md with your changes?
- [x] Have you checked our [Contributing](https://github.com/hdmf-dev/hdmf/blob/dev/docs/CONTRIBUTING.rst) document?
- [x] Have you ensured the PR clearly describes the problem and the solution?
- [x] Is your contribution compliant with our coding style? This can be checked running `flake8` from the source directory.
- [x] Have you checked to ensure that there aren't other open [Pull Requests](https://github.com/hdmf-dev/hdmf/pulls) for the same change?
- [x] Have you included the relevant issue number using "Fix #XXX" notation where XXX is the issue number? By including "Fix #XXX" you allow GitHub to close issue #XXX when the PR is merged.

### CODE REVIEW

**Comment by dsleiter** [comment]
I marked the PR as WIP because there are a few questions to resolve, and I wanted to get some feedback from you all if this is on the right track or not.

**Comment by rly** [comment]
This is great. I was diving into the `GroupValidator` code earlier today to try to handle #525 and I ended up refactoring a few things, but did not clean up as much as you did (I'll hold off until this is merged). 

On first pass, the changes all look good to me. I'll take a deeper look tomorrow.

**Comment by dsleiter** [comment]
Sounds good @rly, sorry this one took so long to complete! I expect to be able to make smaller, more frequent contributions going forwards.

**Review by rly — APPROVED** [review]

**Comment by rly** [comment]
This looks great to me. Can you create a new issue ticket documenting the edge cases where the validator does not work correctly. For example in your last test case that is skipped, when a builder can match multiple specs, it matches against the first spec rather than the most specific spec (in terms of type hierarchy). And if there are any others that you thought of while going through this. Thanks for all your work on this!

**Comment by dsleiter** [comment]
Thanks @rly, glad it looks good!

Here are some new issues based on the discussion here:
* https://github.com/hdmf-dev/hdmf/issues/528
* https://github.com/hdmf-dev/hdmf/issues/529
* https://github.com/hdmf-dev/hdmf/issues/530
* https://github.com/hdmf-dev/hdmf/issues/531

There are a few items from my code comments above where I'm not sure if an issue is required:
1. should we ideally use docval on all classes no matter what, or is it fine to leave it off of module-level utility classes that are not intended or likely to be used outside of the module (eg. `SpecMatcher` and `SpecMatches` in the validator)? Happy to add docval if that is ideal.
2. is the new property `data_type` on `BaseStorageSpec` something that fits well conceptually with hdmf and the schema, or is it just something that has utility within the codebase? If so, then we could add a `neurodata_type` property to nwb specs, and add a clarification in the `hdmf ` documentation about `data_type` and its relationship to `data_type_def` and `data_type_inc`. In addition, it might be beneficial to refactor out the multiple checks for `data_type_def` and `data_type_inc` throughout the code, although that doesn't make any functional improvements.

**Comment by oruebel** [comment]
>     1. should we ideally use docval on all classes no matter what

Use of docval is primarily required when type-validation is important, e.g. for Container classes that are relevant for I/O. Use of docval can also help to ensure consisting formatting of doc-strings, but if type validation is not needed, then just using regular doc-strings is fine (but we should make sure the docstrings are formatted correctly). In some rare cases, when a function is called a lot (e.g., 1000s of times) it can be useful to omit docval to avoid the overhead of validation in lieu of performance.

**Comment by dsleiter** [comment]
@oruebel thanks, makes sense. In this case, I don't think performance is a limiting factor here, and I guess there isn't a good reason not to use docval, so I'll go ahead and add it in (probably with a PR for another validator issue, unless you think it's better as a separate issue)

**Comment by oruebel** [comment]
> , unless you think it's better as a separate issue

Doing this as part of another PR is perfectly fine

**Pull request merged**
